File: matiasmasso/Mat.NET/DTO/DTOEdiversaSalesReport.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOEdiversaSalesReport : DTOBaseGuid
    {
        public string Id { get; set; }
        public DateTime Fch { get; set; }
        public DTOCustomer Customer { get; set; }
        public DTOCur Cur { get; set; }
        public List<Item> Items { get; set; }
        public List<DTOEdiversaException> Exceptions { get; set; }


        public DTOEdiversaSalesReport() : base()
        {
            Items = new List<Item>();
            Exceptions = new List<DTOEdiversaException>();
        }

        public DTOEdiversaSalesReport(Guid oGuid) : base(oGuid)
        {
            Items = new List<Item>();
            Exceptions = new List<DTOEdiversaException>();
        }

        public void AddException(DTOEdiversaException.Cods oCod, string sMsg, DTOEdiversaException.TagCods oTagCod = DTOEdiversaException.TagCods.NotSet, DTOBaseGuid oTag = null/* TODO Change to default(_) if this is not a reference type */)
        {
            var oException = DTOEdiversaException.Factory(oCod, oTag, sMsg);
            oException.TagCod = oTagCod;
            Exceptions.Add(oException);
        }

        public DTOAmt amount()
        {
            var eur = Items.Sum(x => x.Eur);
            return DTOAmt.Factory(eur);
        }

        public class Item
        {
            public DTOCustomer Customer { get; set; }
            public DTOProductSku Sku { get; set; }

            public DateTime Fch { get; set; }
            public string Dept { get; set; }
            public string Centro { get; set; }
            public int Qty { get; set; }
            public int QtyBack { get; set; }

            public decimal Eur { get; set; }
            public decimal Retail { get; set; }
        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOEdiversaSalesReport.vb
Public Class DTOEdiversaSalesReport
    Inherits DTOBaseGuid
    Property Id As String
    Property Fch As Date
    Property Customer As DTOCustomer
    Property Cur As DTOCur
    Property Items As List(Of Item)
    Property Exceptions As List(Of DTOEdiversaException)


    Public Sub New()
        MyBase.New()
        _Items = New List(Of Item)
        _Exceptions = New List(Of DTOEdiversaException)
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
        _Items = New List(Of Item)
        _Exceptions = New List(Of DTOEdiversaException)
    End Sub

    Public Sub AddException(oCod As DTOEdiversaException.Cods, sMsg As String, Optional oTagCod As DTOEdiversaException.TagCods = DTOEdiversaException.TagCods.NotSet, Optional oTag As DTOBaseGuid = Nothing)
        Dim oException = DTOEdiversaException.Factory(oCod, oTag, sMsg)
        oException.TagCod = oTagCod
        _Exceptions.Add(oException)
    End Sub

    Public Function amount() As DTOAmt
        Dim eur = _Items.Sum(Function(x) x.Eur)
        Return DTOAmt.Factory(eur)
    End Function

    Public Class Item

        Property customer As DTOCustomer
        Property sku As DTOProductSku

        Property fch As Date
        Property dept As String
        Property centro As String
        Property qty As Integer
        Property qtyBack As Integer

        Property eur As Decimal
        Property retail As Decimal

    End Class
End Class
